style(*): grammar, spelling, punctuation (#539)
discordjs/guide: guide/voice/understanding-voice.md

<branch version="11.x">

> The Discord.js voice system allows your bot to join voice channels and play audio. This guide will teach you how to make simple music bots and tips and tricks to optimize performance!

Reviewer: ```suggestion
The Discord.js voice system allows your bot to join voice channels and play audio. This guide will teach you how to make simple music bots and give you tips to optimize performance!
```

### Playing audio

To play audio, we need to send Opus audio packets to Discord at a fixed interval–we have selected 20ms. This is the `StreamDispatcher`s job–it is a WritableStream with Opus audio packets written to it. The dispatcher handles the packets' timing and applies some metadata, e.g., the sequence number of the packet, the timestamp, and then encrypts the packet before sending it via the UDP socket.

Obtaining the Opus audio is mainly offloaded to a separate module, [`prism-media`](https://github.com/amishshah/prism-media). This module converts media to Opus audio, either through "demuxing" Ogg/WebM media files (extracting Opus audio directly) or using FFmpeg to transcode most media formats to Opus audio so we can use it.

It can also change the volume of a stream in real-time without much cost to performance.

In conclusion:

1. You do `VoiceConnection#play(...)`
2. Discord.js will convert whatever you've provided into Opus audio in the most efficient way it can find. The worst-case route would be having FFmpeg convert a file to PCM, passing this data through an Opus encoder.

> 3. A `StreamDispatcher` is created, and it is fed data from the Opus audio stream

Reviewer: ```suggestion
3. A `StreamDispatcher` is created and fed data from the Opus audio stream
```

### Key takeaways

- You should be familiar with the definitions in the glossary above
- A voice connection does not disrupt the bot's main gateway–separate connections to a voice gateway and UDP socket are created
- The `StreamDispatcher` is essentially a WritableStream that takes Opus audio
- Receiving audio is officially unsupported; however, we do our best to maintain it

> - Playing audio has varying complexity depending on the file/audio you want to play–the simplest (and most efficient) files would be Ogg/WebM files as they already contain Opus audio. FFmpeg and an Opus Encoder must pass other files before they can play

Reviewer: ```suggestion
- Playing audio has varying complexity depending on the file/audio you want to play–the simplest (and most efficient) files would be Ogg/WebM files as they already contain Opus audio. Other files must be passed through FFmpeg and an Opus Encoder before they can play.
```
